src/cluster_migrate.c

```diff
@@ -0,0 +1,1950 @@
+    /* Since they aren't guaranteed to be sorted, just use a nested loop. */
+    listIter li1, li2;
+    listNode *ln1, *ln2;
+    listRewind(ranges1, &li1);
+    listRewind(ranges2, &li2);
+    while ((ln1 = listNext(&li1)) != NULL) {
+        while ((ln2 = listNext(&li2)) != NULL) {
+            slotRange *range1 = ln1->value;
+            slotRange *range2 = ln2->value;
+            if (doSlotRangesOverlap(range1, range2)) {
+                return 1;
+            }
+        }
+    }
+    return 0;
+}
+
+/* Remove all the keys in the hash slots that are in the given slot range list
+ * and not owned by myself now. */
+void delKeysNotOwnedByMyself(list *slot_ranges) {
+    listNode *ln;
+    listIter li;
+    listRewind(slot_ranges, &li);
+
+    while ((ln = listNext(&li)) != NULL) {
+        slotRange *range = ln->value;
+        for (int i = range->start_slot; i <= range->end_slot; i++) {
+            if (server.cluster->slots[i] != server.cluster->myself) {
+                delKeysInSlot(i, 1, true, false);
+            }
+        }
+    }
+}
+
+void setSlotImportingStateInDb(serverDb *db, list *slot_ranges, int is_importing) {
+    if (db == NULL) return;
+    listNode *ln;
+    listIter li;
+    listRewind(slot_ranges, &li);
+    while ((ln = listNext(&li)) != NULL) {
+        slotRange *range = ln->value;
+        for (int slot = range->start_slot; slot <= range->end_slot; slot++) {
+            kvstoreSetIsImporting(db->keys, slot, is_importing);
+            kvstoreSetIsImporting(db->expires, slot, is_importing);
+        }
+    }
+}
+
+void setSlotImportingStateInAllDbs(list *slot_ranges, int is_importing) {
+    for (int i = 0; i < server.dbnum; i++) {
+        serverDb *db = server.db[i];
+        setSlotImportingStateInDb(db, slot_ranges, is_importing);
+    }
+}
+
+/* Parse as many slot ranges starting from start_index, returning a list of parsed slot ranges, or
+ * NULL if there is an error (in which case err_out will be set). If parsing completes successfully,
+ * end_index_out will be set to the index of the next argument needed to be parsed in c->argv.
+ *
+ * Note that all slots in the slot range should belong to a sole node in the cluster topology.
+ * node_out will be set to the node that owns all the slots. */
+list *parseSlotRangesOrReply(client *c, int start_index, int *end_index_out, clusterNode **node_out) {
+    list *slot_ranges = createSlotRangeList();
+    int startslot, endslot;
+    *node_out = NULL;
+    *end_index_out = c->argc;
+    for (int i = start_index; i < c->argc; i += 2) {
+        if (getLongLongFromObject(c->argv[i], NULL) != C_OK) {
+            /* If we encounter a non-integer parameter, we assume that this is the next argument. */
+            *end_index_out = i;
+            break;
+        }
+        /* Get the current slot range. */
+        if ((startslot = getSlotOrReply(c, c->argv[i])) == -1) {
+            listRelease(slot_ranges);
+            return NULL;
```

Comment:
this pattern is quite tedious error-prone imo. i'd suggest not returning from the middle of the function but rather break out of the loop and then check if we process all the args. if not, free `slot_ranges` just once before returning NULL